pkg/sql/opt/norm/custom_funcs.go

```diff
@@ -2348,7 +2353,7 @@ func (c *CustomFuncs) CastToCollatedString(str opt.ScalarExpr, locale string) op
 	if err != nil {
 		panic(err)
 	}
-	return c.f.ConstructConst(d)
+	return c.f.ConstructConst(d, d.ResolvedType())
```

Comment:
This isn't quite right, since you can have `CollatedString{Char}`, `CollatedString{NVarChar}`, etc, and they're different OIDs. I think should be:
```
return c.f.ConstructConst(d, types.MakeCollated(str.DataType(), locale)
```
This subtlety shows what we're up against. `ResolvedType` can't be relied upon anywhere if OID's matter, unless we're always careful to use `DoidWrapper` (but we're not careful). For us to truly fix the typing, we'd need to audit all of the optbuilder/optimizer code.